app/components/bing-map.js

```diff
@@ -50,47 +59,63 @@ export default Component.extend({
     let el = this.$()[0];
     let opts = get(this, 'mapOptions');
 
-    this.set('map', new Microsoft.Maps.Map(el, opts));
+    let map = new Microsoft.Maps.Map(el, opts);
+    let infoBox = new Microsoft.Maps.Infobox(map.getCenter(), {
+      visible: false
+    });
+    infoBox.setMap(map);
+    this.set('map', map);
+    this.set('infoBox', infoBox);
     this.updateCenter();
+  },
 
+  clearEntities: function(map) {
+    let handlers = get(this, 'handlers');
+    (handlers || []).forEach(  Microsoft.Maps.Events.removeHandler );
+    this.set('handlers', []);
+    map.entities.clear();
+  },
+
+  addPinEvents: function(pin) {
+    let {
+      handlers,
+      events,
+      infoBox,
+      map
+    } = getProperties(this, 'handlers', 'events', 'infoBox', 'map');
+    let pinEvents = getWithDefault(events, 'pin', {});
+
+    Object.keys(pinEvents).map( (eventName) => {
+      handlers.push(Microsoft.Maps.Events.addHandler(
+        pin,
+        eventName,
+        (e) => pinEvents[eventName](e, infoBox, map)
+      ));
+    });
   },
 
   updateCenter: function () {
     let {
       map,
       mapOptions,
-      locations,
-      onPinClick,
-      onPinMouseDown,
-      onPinMouseOver,
-      onPinMouseOut,
-      onPinMouseUp
-    } = getProperties(this, 'map', 'mapOptions', 'locations', 'onPinClick',
-      'onPinMouseDown', 'onPinMouseOver', 'onPinMouseOut', 'onPinMouseUp'
-    );
+      locations
+    } = getProperties(this, 'map', 'mapOptions', 'locations');
 
     if (map) {
       let { center, bounds, zoom, padding } = mapOptions;
       map.setView({ center, bounds, zoom, padding });
-      let infobox = new Microsoft.Maps.Infobox(map.getCenter(), {
-        visible: false
-      });
-      infobox.setMap(map);
-      map.entities.clear();
+      this.clearEntities(map);
       locations.forEach((location) => {
         let pin = new Microsoft.Maps.Pushpin(location.loc, location.options);
         pin.metadata = (location.options || {}).metadata || {};
         map.entities.push(pin)
-        Microsoft.Maps.Events.addHandler(pin, 'click', (e) => onPinClick(e, infobox, map));
-        Microsoft.Maps.Events.addHandler(pin, 'mousedown', (e) => onPinMouseDown(e, infobox, map));
-        Microsoft.Maps.Events.addHandler(pin, 'mouseout', (e) => onPinMouseOut(e, infobox, map));
-        Microsoft.Maps.Events.addHandler(pin, 'mouseover', (e) => onPinMouseOver(e, infobox, map));
-        Microsoft.Maps.Events.addHandler(pin, 'mouseup', (e) => onPinMouseUp(e, infobox, map));
+        this.addPinEvents(pin);
       });
     }
   },
 
   removeMap: function() {
```

Comment:
who calls removeMap?

Reply:
```willDestroyElement() {
    this.removeMap();
  },```

Reply:
willDestroyElement() {
    this.removeMap();
  },

Reply:
perfect. I didn't see that in the diff.